language-plutus-core/test-cost-model/TestCostModels.hs

```diff
@@ -0,0 +1,84 @@
+{-# LANGUAGE DeriveGeneric       #-}
+{-# LANGUAGE FlexibleInstances   #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE QuasiQuotes         #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell     #-}
+{-# LANGUAGE TypeApplications    #-}
+
+import           Control.Applicative
+import           Control.DeepSeq
+import           Control.Monad.Morph
+import           CostModelCreation
+import           Data.Coerce
+import           Data.Functor.Compose
+import           Foreign.R                                          hiding (unsafeCoerce)
+import           GHC.Generics                                       (Generic, Generic1)
+import           H.Prelude                                          (MonadR, Region, r)
+import           Hedgehog
+import qualified Hedgehog.Gen                                       as Gen
+import           Hedgehog.Main
+import qualified Hedgehog.Range                                     as Range
+import           Language.PlutusCore.Evaluation.Machine.ExBudgeting
+import           Language.PlutusCore.Evaluation.Machine.ExMemory
+import           Language.R                                         hiding (unsafeCoerce)
+import           Unsafe.Coerce                                      (unsafeCoerce)
+
+
+prop_addInteger = testPredict addInteger (getConst . paramAddInteger)
+prop_subtractInteger = testPredict subtractInteger (getConst . paramSubtractInteger)
+prop_multiplyInteger = testPredict multiplyInteger (getConst . paramMultiplyInteger)
+prop_divideInteger = testPredict divideInteger (getConst . paramDivideInteger)
+prop_quotientInteger = testPredict quotientInteger (getConst . paramQuotientInteger)
+prop_remainderInteger = testPredict remainderInteger (getConst . paramRemainderInteger)
+prop_modInteger = testPredict modInteger (getConst . paramModInteger)
+prop_lessThanInteger = testPredict lessThanInteger (getConst . paramLessThanInteger)
+prop_greaterThanInteger = testPredict greaterThanInteger (getConst . paramGreaterThanInteger)
+prop_lessThanEqInteger = testPredict lessThanEqInteger (getConst . paramLessThanEqInteger)
+prop_greaterThanEqInteger = testPredict greaterThanEqInteger (getConst . paramGreaterThanEqInteger)
+prop_eqInteger = testPredict eqInteger (getConst . paramEqInteger)
+  -- testPredict concatenate "concatenateModel"
+  -- testPredict takeByteString "takeByteStringModel"
+  -- testPredict dropByteString "dropByteStringModel"
+  -- testPredict sHA2 "sha2Model"
+  -- testPredict sHA3 "sha3Model"
+  -- testPredict verifySignature "verifySignatureModel"
+  -- testPredict eqByteString "eqByteStringModel"
+  -- testPredict ltByteString "ltByteStringModel"
+  -- testPredict gtByteString "gtByteStringModel"
+  -- testPredict ifThenElse "ifThenElseModel"
+
+probablySafeHoist :: (MFunctor t, Monad m) => (m () -> n ()) -> t m () -> t n ()
+probablySafeHoist nt = hoist (unsafeCoerce nt)
```

Comment:
:thinking:

Reply:
It works, doesn't segfault, and I tested it for the full domain. So it could even be `safeHoist` :grin:.

Reply:
This assumes that `t m a` always contains an `m a` (or multiple of them), but this goes against the semantics of the class, because otherwise the type of `hoist` would be 

```haskell
hoist :: (m a -> n a) -> t m a -> t n a
```

It's no surprise `hoist` is more general: it's designed to support `pipes` (as both the packages are written by the same author) and there `m` holds [something that is not just `a`](http://hackage.haskell.org/package/pipes-4.3.13/docs/src/Pipes.Internal.html#M). Same applies to `Streaming` for another example.

So I think it should be called `unsafeHoist`.

Reply:
Got some help from #haskell, it breaks on `const (pure ())` passed as function, so yeah, it's `unsafe`.

Reply:
Left a longer comment on why it probably works and that you get to keep the pieces if to breaks.